---
title: "Differential gene expression analysis in HFpEF patients"
author: "Hildur"
date: "January, 2020"
output: word_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=TRUE, message = F)
suppressMessages(library(DESeq2))
suppressMessages(library(limma))
suppressMessages(library(data.table))
suppressMessages(library(Heatplus))
suppressMessages(library(readxl))
suppressMessages(library(BiocParallel))
register(MulticoreParam(4))
suppressMessages(library(ggplot2))
suppressMessages(library(ggrepel))
suppressMessages(library(dplyr))
suppressMessages(library(AnnotationDbi))
suppressMessages(library(org.Hs.eg.db))
```

## Pre-processing

The raw data were aligned and sorted using hisat2 and counts were estimated using HTseq-count.


## Differential Expression

Differential expression will be estimated using DESeq2.

Setup results folder and file naming.
```{r}
useFolder = "Results/"
figFolder = "Figures/"
name_suffix = "_hf_deg_v1"
filter_reads = 50
```



### Design matrix for analysis

Differential expression between the control samples and HFpEF samples was calculated by reading in the counts generated by HTSeq.

```{r}
samplesInfo = read_excel("../VirginiaHahn_dataTable.xlsx")
samplesInfo <- as.data.frame(samplesInfo)
samplesTable = data.frame(samplenames=samplesInfo$SequenceRunId,
                  samplefiles=samplesInfo$FileName,
                  phenotype = samplesInfo$Group,
                  tissue = samplesInfo$`Tissue Abbr`,
                  disease = samplesInfo$`Disease Abbr`,
                  sex = samplesInfo$`Subject Sex`,
                  age = samplesInfo$`Subject Approximate Age`)
samplesTable$disease<-factor(samplesTable$disease)
samplesTable$disease <- relevel(x = samplesTable$disease, ref = "Normal")
subSamples <- samplesTable[samplesTable$tissue == "RVS",]
```


```{r, echo=TRUE}
setwd("../")
dds <- DESeqDataSetFromHTSeqCount(samplesTable, directory = "counts/", design = ~ tissue + disease)
dds_subset <- DESeqDataSetFromHTSeqCount(subSamples, directory = "counts/", design = ~ disease)
```

###Save table with sample information
```{r}
knitr::kable(samplesTable)
write.csv(samplesTable[,c("samplenames","disease","tissue")],row.names =FALSE,paste0(useFolder,"SamplesTable",name_suffix,".csv"))
```

### Save files for submission
First all the raw reads
```{r}
RawReadsMatrix<-counts(dds)
dim(RawReadsMatrix)

write.csv(RawReadsMatrix,row.names =TRUE,paste0(useFolder,"RawReads_ALLgenes",name_suffix,".csv"))

openxlsx::write.xlsx(RawReadsMatrix, file = paste0(useFolder,"RawReads_ALLgenes",name_suffix,".xlsx"), 
                     sheetName = "Reads", overwrite=TRUE)
```

Split RVS and LVS for file size restriction
```{r}
RawLVS<-RawReadsMatrix[,colnames(RawReadsMatrix)%in%samplesTable[samplesTable$tissue=="LVS","samplenames"]]
RawRVS<-RawReadsMatrix[,colnames(RawReadsMatrix)%in%samplesTable[samplesTable$tissue=="RVS","samplenames"]]
dim(RawReadsMatrix)
dim(RawRVS)
dim(RawLVS)

openxlsx::write.xlsx(RawRVS, file = paste0(useFolder,"RawReads_RVS_ALLgenes",name_suffix,".xlsx"), 
                     sheetName = "Reads", overwrite=TRUE)
openxlsx::write.xlsx(RawLVS, file = paste0(useFolder,"RawReads_LVS_ALLgenes",name_suffix,".xlsx"), 
                     sheetName = "Reads", overwrite=TRUE)
```


###Exclude red blood cell genes
Read a list of genes that should be excluded in all down-stream analysis. These are genes highly (above 0.5) correlated with the expression of HBB and documented as expressed in cells that are precursors to red blood cells.
```{r rbc-genes}
excludeGenes <-read_excel("/Users/hildur/Dropbox/Virginia_Hildur/ExcludeRBCgenes.xlsx",sheet = 1)
dds <- dds[!rownames(dds) %in% excludeGenes$Ensembl,]
dds_subset<- dds_subset[!rownames(dds_subset) %in% excludeGenes$Ensembl,]
```


DESeq2 estimates the sample sizes (sequencing depth per sample), dispersion (variance in read counts for genes for the negative binomial model), automatically to compute the differential expression fold changes and p-values.

Before we will filter out low count reads, based on median expression in each group.
```{r}
sampnr <- dim(subSamples)[1]
dds <- estimateSizeFactors(dds)
dds_subset <- estimateSizeFactors(dds_subset)
keepPEF <-rowMedians(counts(dds[,colnames(dds) %in% 
                             samplesTable[samplesTable$disease=="HFpEF",]$samplenames ],
                       normalized=TRUE))>= filter_reads
keepREF <-rowMedians(counts(dds[,colnames(dds) %in% 
                             samplesTable[samplesTable$disease=="HFrEF",]$samplenames ],
                       normalized=TRUE))>= filter_reads
keepCTRL <-rowMedians(counts(dds[,colnames(dds) %in% 
                             samplesTable[samplesTable$disease=="Normal",]$samplenames ],
                       normalized=TRUE))>= filter_reads
keep <- rowSums(cbind(keepPEF,keepREF,keepCTRL)) >0
dds <- dds[keep,]

keepsubPEF <-rowMedians(counts(dds_subset[,colnames(dds_subset) %in% 
                             subSamples[subSamples$disease=="HFpEF",]$samplenames ],
                       normalized=TRUE))>= filter_reads
keepsubREF <-rowMedians(counts(dds_subset[,colnames(dds_subset) %in% 
                             subSamples[subSamples$disease=="HFrEF",]$samplenames ],
                       normalized=TRUE))>= filter_reads
keepsubCTRL <-rowMedians(counts(dds_subset[,colnames(dds_subset) %in% 
                             subSamples[subSamples$disease=="Normal",]$samplenames ],
                       normalized=TRUE))>= filter_reads
keep_subset <- rowSums(cbind(keepsubPEF,keepsubREF,keepsubCTRL)) >0
dds_subset <- dds_subset[keep_subset,]
```



```{r, echo=TRUE}
dds <- DESeq(dds,parallel=TRUE, BPPARAM=MulticoreParam(8)) 
dds_subset <- DESeq(dds_subset,parallel=TRUE, BPPARAM=MulticoreParam(8))

resultsNames(dds)
resultsNames(dds_subset)

```


```{r}
cook = TRUE
results.rvslvs.hfpef = results(dds, cooksCutoff=cook,tidy = TRUE, contrast = c("disease","HFpEF","Normal"), alpha=0.05)
results.rvslvs.hfref = results(dds, cooksCutoff=cook,tidy = TRUE, contrast = c("disease","HFrEF","Normal"), alpha=0.05)
results.rvslvs.pefref = results(dds, cooksCutoff=cook,tidy = TRUE, contrast = c("disease","HFpEF","HFrEF"), alpha=0.05)

results.rvs.pef <- results(dds_subset, cooksCutoff=cook,tidy = TRUE, contrast = c("disease","HFpEF","Normal"), alpha=0.05)
results.rvs.ref <- results(dds_subset, cooksCutoff=cook,tidy = TRUE, contrast = c("disease","HFrEF","Normal"), alpha=0.05)
results.rvs.pefref <- results(dds_subset, cooksCutoff=cook,tidy = TRUE, contrast = c("disease","HFpEF","HFrEF"), alpha=0.05)


normalID = samplesTable[samplesTable$disease =="Normal","samplenames"]
hfpefID = samplesTable[samplesTable$disease =="HFpEF","samplenames"]
hfrefID = samplesTable[samplesTable$disease =="HFrEF","samplenames"]

MedianColumn <- function(data,samples){
  x=rowMedians(assays(data)[["replaceCounts"]][,colnames(counts(data)) %in% samples])
  return(x)
}

results.rvs.pef$Median.reads = rowMedians(assays(dds_subset)[["replaceCounts"]])
results.rvs.pef$control_median.reads = MedianColumn(dds_subset,normalID)
results.rvs.pef$hfpef_median.reads = MedianColumn(dds_subset,hfpefID)
results.rvs.pef$hfref_median.reads = MedianColumn(dds_subset,hfrefID)

results.rvslvs.hfpef$Median.reads = rowMedians(assays(dds)[["replaceCounts"]])
results.rvslvs.hfpef$control_median.reads = MedianColumn(dds,normalID)
results.rvslvs.hfpef$hfpef_median.reads = MedianColumn(dds,hfpefID)
results.rvslvs.hfpef$hfref_median.reads = MedianColumn(dds,hfrefID)
```

#Data quality
Have a look at data distributions and outliers 
```{r}
#Flag genes with high variance. Use the Cook's distance

par(mar=c(8,5,2,2))
boxplot(log10(assays(dds)[["cooks"]]), range=0, las=2)

#Dispersion
plotDispEsts(dds_subset)
plotDispEsts(dds)
hist(mcols(dds)$dispGeneEst, breaks=500, ylim = c(0,100))
sum(mcols(dds)$dispGeneEst > 1)


resGA <- results(dds_subset, lfcThreshold=0.01, altHypothesis="greaterAbs")
plotMA(resGA)
W <- results.rvs.pef$stat
maxCooks <- apply(assays(dds_subset)[["cooks"]],1,max)
idx <- !is.na(W)
plot(rank(W[idx]), maxCooks[idx], xlab="rank of Wald statistic", 
     ylab="maximum Cook's distance per gene",
     ylim=c(0,5), cex=.4, col=rgb(0,0,0,.3))
m <- ncol(dds)
p <- 3
abline(h=qf(.99, p, m - p))
```


###Combine data in data frame
Create two lists for RVS only (95 samples) and for the RVS and LVS combined (145 samples)
```{r put-all-results-together}

allresultsRVS <- data.frame(geneid = results.rvs.pef$row, 
                         pvalue.pef = results.rvs.pef$pvalue,
                         pvalue.ref = results.rvs.ref$pvalue,
                         pvalue.pef.ref = results.rvs.pefref$pvalue,
                         padj.pef = results.rvs.pef$padj,
                         padj.ref = results.rvs.ref$padj,
                         padj.pef.ref = results.rvs.pefref$padj,
                         l2fc.pef = results.rvs.pef$log2FoldChange,
                         l2fc.ref = results.rvs.ref$log2FoldChange,
                         l2fc.pef.ref = results.rvs.pefref$log2FoldChange,
                         control_median = results.rvs.pef$control_median.reads,
                         hfpef_median = results.rvs.pef$hfpef_median.reads,
                         hfref_median = results.rvs.pef$hfref_median.reads,
                         stringsAsFactors = F
                         )
allresultsRVS <- as.data.table(allresultsRVS)
head(allresultsRVS)

allresultsRVSLVS <- data.frame(geneid = results.rvslvs.hfpef$row, 
                         pvalue.pef = results.rvslvs.hfpef$pvalue,
                         pvalue.ref = results.rvslvs.hfref$pvalue,
                         pvalue.pef.ref = results.rvslvs.pefref$pvalue,
                         padj.pef = results.rvslvs.hfpef$padj,
                         padj.ref = results.rvslvs.hfref$padj,
                         padj.pef.ref = results.rvslvs.pefref$padj,
                         l2fc.pef = results.rvslvs.hfpef$log2FoldChange,
                         l2fc.ref = results.rvslvs.hfref$log2FoldChange,
                         l2fc.pef.ref = results.rvslvs.pefref$log2FoldChange,
                         control_median = results.rvslvs.hfpef$control_median.reads,
                         hfpef_median = results.rvslvs.hfpef$hfpef_median.reads,
                         hfref_median = results.rvslvs.hfpef$hfref_median.reads,
                         stringsAsFactors = F
                         )
allresultsRVSLVS <- as.data.table(allresultsRVSLVS)
head(allresultsRVSLVS)
```

Z-scores were calculated for each gene based on the p-value of differential expression. z-scores were defined as abs(qnorm(pval/2)) * sign(log2FoldChange). For all genes with zero read counts or those that DESeq2 filtered out and a p-value of fold change estimated was not calculated, the z-score is zero.

```{r calculate-z-score}
allresultsRVS <-allresultsRVS[, z.pos.pef := abs(qnorm(pvalue.pef/2)) * sign(l2fc.pef)]
allresultsRVS <-allresultsRVS[, z.pos.ref := abs(qnorm(pvalue.ref/2)) * sign(l2fc.ref)]

allresultsRVSLVS <-allresultsRVSLVS[, z.pos.pef := abs(qnorm(pvalue.pef/2)) * sign(l2fc.pef)]
allresultsRVSLVS <-allresultsRVSLVS[, z.pos.ref := abs(qnorm(pvalue.ref/2)) * sign(l2fc.ref)]
```


####Ensemble to gene names

```{r ensembl-hugo-conversion}
library(biomaRt)
human = useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")
#when the server is not responding use this mirror
#human = useEnsembl("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", mirror = "useast")
hggenes <- getBM(attributes = c("ensembl_gene_id","external_gene_name","description","entrezgene_id"),
                  filters = "ensembl_gene_id", values = allresultsRVS$geneid,
                 mart = human, uniqueRows=T)
hggenes <- as.data.table(hggenes)
setkey(hggenes, external_gene_name)


hggenes.RVSLVS <- getBM(attributes = c("ensembl_gene_id","external_gene_name","description","entrezgene_id"),
                  filters = "ensembl_gene_id", values = allresultsRVSLVS$geneid,
                 mart = human, uniqueRows=T)
hggenes.RVSLVS <- as.data.table(hggenes)
setkey(hggenes.RVSLVS, external_gene_name)
```
Merge with gene name
```{r,echo=FALSE}
setkey(allresultsRVS, geneid)
allresultsRVSnames <- merge(x = allresultsRVS, y = hggenes, by.x="geneid", by.y="ensembl_gene_id", all.x=TRUE)
allresultsRVSnames <- allresultsRVSnames[,c(1,16,18,2:15,17)]

allresultsRVSLVSnames <- merge(x = allresultsRVSLVS, y = hggenes, by.x="geneid", by.y="ensembl_gene_id", all.x=TRUE)
allresultsRVSLVSnames <- allresultsRVSLVSnames[,c(1,16,18,2:15,17)]
```


Save to file all results and the z-score file

```{r save-file}
allresultsRVSnames <- allresultsRVSnames[order(allresultsRVSnames$pvalue.pef, decreasing = F),]
openxlsx::write.xlsx(allresultsRVSnames, file = paste0(useFolder,"allresultsRVSnames_manuscript",name_suffix,".xlsx"), 
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)

allresultsRVSLVSnames <- allresultsRVSLVSnames[order(allresultsRVSLVSnames$pvalue.pef, decreasing = F),]
openxlsx::write.xlsx(allresultsRVSLVSnames, file = paste0(useFolder,"allresultsRVSLVSnames_manuscript",name_suffix,".xlsx"), 
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)

#How many DEgenes
table(allresultsRVSnames$pvalue.pef<1e-6)
table(allresultsRVSLVSnames$pvalue.pef<1e-6)
```


```{r write-z-scores-KO}
FWER <- 1/dim(allresultsRVSLVSnames)[1]
allSigResults1 <- subset(allresultsRVSLVSnames, pvalue.pef <= FWER)
allSigResults <- subset(allSigResults1, abs(l2fc.pef) >0.0)
write.table(x = allSigResults[! is.na(external_gene_name), .(external_gene_name, z.pos.pef)],
            file = paste0(useFolder,"pos.zscores.txt"),
            row.names = F,
            col.names = c("gene", "zscore"),
            quote = F,
            sep = "\t")
```

Write vst transformed data for WGCNA analysis. RVs only and HFpEF only
```{r write-vst-genes}
dds_vst = vst(assays(dds)[["replaceCounts"]])
dds_subset_vst = vst(assays(dds_subset)[["replaceCounts"]])
MyListRaw = as.data.frame(dds_subset_vst)
MyListRaw$geneid <- rownames(MyListRaw)
MyListRawName <- merge(x = MyListRaw, y = hggenes, by.x="geneid", by.y="ensembl_gene_id", all.x=TRUE)
MyListRawName <- MyListRawName[complete.cases(MyListRawName$external_gene_name),]
idxRaw = duplicated(MyListRawName$external_gene_name)
MyListRawName <- MyListRawName[!idxRaw,]
rownames(MyListRawName) <- MyListRawName$external_gene_name
MyListRawName  <- MyListRawName[,-which(colnames(MyListRawName) %in% c("geneid","hgnc_symbol","external_gene_name","description","Chromosome.scaffold.name","entrezgene_id"))]
MyListRawName<-MyListRawName[,as.numeric(colnames(MyListRawName)) %in% hfpefID]

openxlsx::write.xlsx(MyListRawName, file =paste0(useFolder,"HFpEFGenesVSTWithGeneName_RVS",name_suffix,".xlsx"), 
                     sheetName = "GenesVST", overwrite=TRUE,
                     asTable = T)
```


Save the R environment to reload for downstream analysis
```{r echo=FALSE}
save.image(file='HF_DEGanalysis.RData')
#load('HF_DEGanalysis.RData')
```




